contracts/Staking.sol

```diff
@@ -0,0 +1,559 @@
+  mapping(address => uint256) undelegated; // delegator => totalUndelegated
+  mapping(address => mapping(address => mapping(address => uint256))) pendingRedelegateTime; // delegator => srcValidator => dstValidator => minTime
+
+  bool internal locked;
+
+  modifier noReentrant() {
+    require(!locked, "No re-entrancy");
+    locked = true;
+    _;
+    locked = false;
+  }
+
+  modifier tenDecimalPrecision(uint256 amount) {
+    require(msg.value%TEN_DECIMALS==0 && amount%TEN_DECIMALS==0, "precision loss in conversion");
+    _;
+  }
+
+  modifier initParams() {
+    if (!alreadyInit) {
+      oracleRelayerFee = INIT_ORACLE_RELAYER_FEE;
+      minDelegation = INIT_MIN_DELEGATION;
+      alreadyInit = true;
+    }
+    _;
+  }
+
+  /*********************************** Events **********************************/
+  event delegateSubmitted(address indexed delegator, address indexed validator, uint256 amount, uint256 oracleRelayerFee);
+  event undelegateSubmitted(address indexed delegator, address indexed validator, uint256 amount, uint256 oracleRelayerFee);
+  event redelegateSubmitted(address indexed delegator, address indexed validatorSrc, address indexed validatorDst, uint256 amount, uint256 oracleRelayerFee);
+  event rewardReceived(address indexed delegator, uint256 amount);
+  event rewardClaimed(address indexed delegator, uint256 amount);
+  event undelegatedReceived(address indexed delegator, address indexed validator, uint256 amount);
+  event undelegatedClaimed(address indexed delegator, uint256 amount);
+  event failedDelegate(address indexed delegator, address indexed validator, uint256 amount, uint8 errCode);
+  event failedUndelegate(address indexed delegator, address indexed validator, uint256 amount, uint8 errCode);
+  event failedRedelegate(address indexed delegator, address indexed valSrc, address indexed valDst, uint256 amount, uint8 errCode);
+  event paramChange(string key, bytes value);
+  event failedSynPackage(uint8 indexed eventCode, uint256 errCode);
+  event crashResponse(uint8 indexed eventCode);
+
+  receive() external payable {}
+
+  /************************* Implement cross chain app *************************/
+  function handleSynPackage(uint8, bytes calldata msgBytes) external onlyCrossChainContract initParams override returns(bytes memory) {
+    RLPDecode.Iterator memory iter = msgBytes.toRLPItem().iterator();
+    uint8 eventCode = uint8(iter.next().toUint());
+    uint32 resCode;
+    if (eventCode == EVENT_DISTRIBUTE_REWARD) {
+      resCode = _handleDistributeRewardSynPackage(iter);
+    } else if (eventCode == EVENT_DISTRIBUTE_UNDELEGATED) {
+      resCode = _handleDistributeUndelegatedSynPackage(iter);
+    } else {
+      resCode = ERROR_UNKNOWN_PACKAGE_TYPE;
+    }
+
+    if (resCode == CODE_OK) {
+      return new bytes(0);
+    } else {
+      emit failedSynPackage(eventCode, resCode);
+      return _RLPEncode(eventCode, msgBytes);
```

Comment:
It will contains two `eventCode` in the return message, it seems not our expectation.

Reply:
Fixed